File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution
{
public:
    void dfs(TreeNode *root, int max, int &res)
    {
        if (root == nullptr)
            return;
        if(root->val >= max) {
            max = root->val;
            res++;
        }
        dfs(root->left, max, res);
        dfs(root->right, max, res);
    }

    int goodNodes(TreeNode *root)
    {
        int res = 0;
        dfs(root, -101, res);
        return res;
    }
};

File: Data Structures & Algorithms/binary-tree-diameter/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:

    dia = 0

    def f(self, root):
        if(not root):
            return 0
        leftHeight = self.f(root.left)
        rightHeight = self.f(root.right)
        self.dia = max(self.dia, leftHeight + rightHeight + 1)        
        return 1 + max(leftHeight, rightHeight)

    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.f(root)
        return self.dia - 1


        